components/camel-etcd/src/test/java/org/apache/camel/component/etcd/support/EtcdTestSupport.java

```diff
@@ -0,0 +1,95 @@
+import mousio.etcd4j.EtcdClient;
+import mousio.etcd4j.responses.EtcdKeysResponse;
+import org.apache.camel.CamelContext;
+import org.apache.camel.Processor;
+import org.apache.camel.component.etcd.EtcdHelper;
+import org.apache.camel.component.etcd.EtcdKeysComponent;
+import org.apache.camel.component.etcd.EtcdStatsComponent;
+import org.apache.camel.component.etcd.EtcdWatchComponent;
+import org.apache.camel.test.testcontainers.junit5.ContainerAwareTestSupport;
+import org.apache.camel.test.testcontainers.junit5.Wait;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.GenericContainer;
+
+public abstract class EtcdTestSupport extends ContainerAwareTestSupport {
+    public static final ObjectMapper MAPPER = EtcdHelper.createObjectMapper();
+    public static final String CONTAINER_IMAGE = "quay.io/coreos/etcd:v2.3.7";
+    public static final String CONTAINER_NAME = "etcd";
+    public static final int ETCD_CLIENT_PORT = 2379;
+    public static final int ETCD_PEER_PORT = 2380;
+
+    protected static final Processor NODE_TO_VALUE_IN = exchange -> {
+        EtcdKeysResponse response = exchange.getMessage().getBody(EtcdKeysResponse.class);
+        if (response != null) {
+            exchange.getMessage().setBody(response.node.key + "=" + response.node.value);
+        }
+    };
+
+    protected String getClientUri() {
+        return String.format(
+            "http://%s:%d",
+            getContainerHost(CONTAINER_NAME),
+            getContainerPort(CONTAINER_NAME, ETCD_CLIENT_PORT));
+    }
+
+    protected EtcdClient getClient() {
+        return new EtcdClient(URI.create(getClientUri()));
+    }
+
+    @Override
+    protected CamelContext createCamelContext() throws Exception {
+        EtcdKeysComponent keys = new EtcdKeysComponent();
+        keys.setUris(getClientUri());
+
+        EtcdStatsComponent stats = new EtcdStatsComponent();
+        stats.setUris(getClientUri());
+
+        EtcdWatchComponent watch = new EtcdWatchComponent();
+        watch.setUris(getClientUri());
+
+        CamelContext context = super.createCamelContext();
+        context.addComponent("etcd-keys", keys);
+        context.addComponent("etcd-stats", stats);
+        context.addComponent("etcd-watch", watch);
+
+        return context;
+    }
+
+    @Override
+    protected GenericContainer<?> createContainer() {
+        return etcdContainer();
+    }
+
+    public static GenericContainer etcdContainer() {
+        return new GenericContainer(CONTAINER_IMAGE)
+            .withNetworkAliases(CONTAINER_NAME)
+            .withExposedPorts(ETCD_CLIENT_PORT, ETCD_PEER_PORT)
+            .waitingFor(Wait.forLogMessageContaining("etcdserver: set the initial cluster version", 1))
+            .withCommand(
+                "-name", CONTAINER_NAME + "-0",
+                "-advertise-client-urls", "http://" + DockerClientFactory.instance().dockerHostIpAddress() + ":" + ETCD_CLIENT_PORT,
```

Comment:
actually, I just realized that `ETCD_CLIENT_PORT` will not be the same as the randomly selected port.
Since it works, perhaps you don't need it at all?

Reply:
seems to fail top start if that option is not set

Reply:
just out of curiosity - can you try using a non-default port?
I have a guess that it reconnects to localhost where you already have etcd running on the default port.

Reply:
When I look at the [doc](https://etcd.io/docs/v3.4.0/op-guide/container/) it looks like `advertise-client-urls` and `listen-client-urls` are both bound to the same port

Reply:
this is because they use static ports. It reminds me of Kafka, he had to fight with the same issue there, this is why we have a bit fancy implementation of KafkaContainer :D